dvc/repo/data.py

```diff
@@ -0,0 +1,246 @@
+    if new_obj is None:
+        return {"deleted": [root], **d}
+    if old_obj.hash_info != new_obj.hash_info:
+        return {"modified": [root], **d}
+    return {"unchanged": [root], **d}
+
+
+def granular_diff(
+    root: str,
+    old_obj: Optional["HashFile"],
+    new_obj: Optional["HashFile"],
+    cache: "HashFileDB",
+    with_dirs: bool = False,
+) -> Dict[str, List[str]]:
+    from dvc_data.diff import ROOT
+    from dvc_data.diff import diff as odiff
+    from dvc_data.objects.tree import Tree
+
+    def path_join(root: str, *paths: str) -> str:
+        if not isinstance(new_obj, Tree):
+            return root
+        return os.path.sep.join([root, *paths])
+
+    diff_data = odiff(old_obj, new_obj, cache)
+    drop_root = not with_dirs and isinstance(new_obj, Tree)
+
+    output: Dict[str, List[str]] = defaultdict(list)
+    for state in ("added", "deleted", "modified", "unchanged"):
+        items = getattr(diff_data, state)
+        output[state].extend(
+            path_join(root, *item.new.key)
+            for item in items
+            if not (drop_root and item.new.key == ROOT)
+        )
+        # TODO: PERF: diff is checking not_in_cache for each even if we only
+        # need it for the index.
+        # BUG: not_in_cache file also shows up as modified in staged and
+        # unstaged. We currently don't know if it is really modified.
+        output["not_in_cache"].extend(
+            path_join(root, *item.new.key)
+            for item in items
+            if not item.old.in_cache
+            and not (drop_root and item.new.key == ROOT)
+            and state != "added"
+        )
+    return output
+
+
+def diff(
+    root: str,
+    old_obj: Optional["HashFile"],
+    new_obj: Optional["HashFile"],
+    cache: "HashFileDB",
+    granular: bool = False,
+    with_dirs: bool = False,
+) -> Dict[str, List[str]]:
+    if granular:
+        return granular_diff(
+            root, old_obj, new_obj, cache, with_dirs=with_dirs
+        )
+    return shallow_diff(root, old_obj, new_obj, cache)
+
+
+class GitInfo(TypedDict, total=False):
+    staged: Dict[str, List[str]]
+    unstaged: Dict[str, List[str]]
+    untracked: List[str]
+    is_empty: bool
+    is_dirty: bool
+
+
+def git_info(scm: "Base", untracked_files: str = "all") -> GitInfo:
+    from scmrepo.exceptions import SCMError
+
+    from dvc.scm import NoSCM
+
+    if isinstance(scm, NoSCM):
+        return {}
+
+    try:
+        scm.get_rev()
+    except SCMError:
+        empty_repo = True
+    else:
+        empty_repo = False
+
+    staged, unstaged, untracked = scm.status(untracked_files=untracked_files)
+    # NOTE: order is important here.
+
+    return GitInfo(
+        staged=staged,
+        unstaged=unstaged,
+        untracked=untracked,
+        is_empty=empty_repo,
+        is_dirty=any([staged, unstaged, untracked]),
+    )
+
+
+def diff_index_to_wtree(repo: "Repo", **kwargs: Any) -> Dict[str, List[str]]:
```

Comment:
Let's make these hidden
```suggestion
def _diff_index_to_wtree(repo: "Repo", **kwargs: Any) -> Dict[str, List[str]]:
```

Reply:
Same with others in this file.